File: constraints/pipelines/darkMatter/haloMassFunctionPostProcess.pl
#!/usr/bin/env perl
use strict;
use warnings;
use lib $ENV{'GALACTICUS_EXEC_PATH'         }."/perl";
use lib $ENV{'GALACTICUS_ANALYSIS_PERL_PATH'}."/perl";
use PDL;
use PDL::IO::HDF5;
use PDL::NiceSlice;
use PDL::Math;
use XML::Simple;
use Data::Dumper;
use GnuPlot::PrettyPlots;
use GnuPlot::LaTeX;
use Galacticus::Options;
use Galacticus::Launch::Hooks;
use Galacticus::Launch::PBS;
use Galacticus::Launch::Slurm;
use Galacticus::Launch::Local;
use Galacticus::Constraints::HaloMassFunctions qw(iterate);

# Generate a halo mass function using the optimal parameters.
# Andrew Benson (22-September-2020)

# Get arguments.
my %options =
    (
     force                 => "no", # If yes, force models to be re-run even if they already exist.
     countParticlesMinimum => 300   # The minimum number of particles per halo to include in constraints.
    );
&Galacticus::Options::Parse_Options(\@ARGV,\%options);
die("no `--pipelinePath` option given")
    unless ( exists($options{'pipelinePath'}) );
die("no `--outputDirectory` option given")
    unless ( exists($options{'outputDirectory'}) );

# Ensure paths are correctly suffixed.
foreach my $path ( 'pipelinePath', 'outputDirectory' ) {
    $options{$path} .= "/"
	unless ( $options{$path} =~ m/\/$/ );
}

# Parse config options.
my $queueManager = &Galacticus::Options::Config(                'queueManager' );
my $queueConfig  = &Galacticus::Options::Config($queueManager->{'manager'     });

# Parse the simulations definition file.
my $xml = new XML::Simple();
my $simulations = $xml->XMLin(
    $options{'pipelinePath'}."haloMassFunctionSimulations.xml",
    ForceArray => [ "suite"         , "group"         , "simulation"          ],
    KeyAttr    => {  suite => "name",  group => "name",  simulation => "name" }
    );

# Determine if models/plots should be forcibly remade.
$options{'force'} = "yes"
    if ( exists($options{'select'}) );

# Construct file names for models.
my @entries = &iterate($simulations,\%options);
foreach my $entry ( @entries ) {
    # Construct the parameter and model file names.
    my $identifier = $entry->{'suite'}->{'name'}."_".$entry->{'group'}->{'name'}."_".$entry->{'simulation'}->{'name'}."_".$entry->{'realization'}."_z".$entry->{'redshift'};
    $entry->{'fileParameters'} = $options{'outputDirectory'}."haloMassFunctionBase_".$identifier.".xml";
    $entry->{'filePrefix'    } = $options{'outputDirectory'}."haloMassFunction_"    .$identifier       ;
}

# Create sets of models that can be plotted together.
my %setsRealization;
foreach my $entry ( @entries ) {
    my $identifier = $entry->{'suite'}->{'name'}."_".$entry->{'group'}->{'name'}."_".$entry->{'simulation'}->{'name'}."_z".$entry->{'redshift'};
    push(
	@{$setsRealization{$identifier}},
	$entry
	);
}
my %identifiersHiRes;
my %setsResolution;
foreach my $entry ( @entries ) {
    next
	unless ( $entry->{'group'}->{'name'} =~ m/:hires$/ );
    (my $identifierLoRes = $entry->{'suite'}->{'name'}."_".$entry->{'group'}->{'name'}."_".$entry->{'simulation'}->{'name'}."_z".$entry->{'redshift'}) =~ s/:hires//;
    ++$identifiersHiRes{$identifierLoRes};
}
foreach my $entry ( @entries ) {
    (my $identifierLoRes = $entry->{'suite'}->{'name'}."_".$entry->{'group'}->{'name'}."_".$entry->{'simulation'}->{'name'}."_z".$entry->{'redshift'}) =~ s/:hires//;
    next
	unless ( exists($identifiersHiRes{$identifierLoRes}) );
    if ( $entry->{'group'}->{'name'} =~ m/:hires$/ ) {
	push(@{$setsResolution{$identifierLoRes}->{'hiRes'}},$entry);
    } else {
    	push(@{$setsResolution{$identifierLoRes}->{'loRes'}},$entry);
    }
}

# Generate jobs to create the models.
my @jobs;
## Jobs using our model.
push(
    @jobs,
    map
    {
	-e $_->{'filePrefix'}.":MPI0000.hdf5" && $options{'force'} eq "no"
	?
        ()
	:
	{
	    command    => $ENV{'GALACTICUS_EXEC_PATH'}."/Galacticus.exe ".$_->{'fileParameters'},
	    launchFile => $_->{'filePrefix'}.".sh" ,
	    logFile    => $_->{'filePrefix'}.".log",
	    label      => "haloMassFunction_".$_->{'suite'}->{'name'}."_".$_->{'group'}->{'name'}."_".$_->{'simulation'}->{'name'}."_".$_->{'realization'}."_z".$_->{'redshift'},
	    ppn        => 1,
	    nodes      => 1,
	    mpi        => "no"
	}
    }
    @entries
    );
## Jobs with no numerical effects.
push(
    @jobs,
    map
    {
	-e $_->{'filePrefix'}.".hdf5_true:MPI0000" && $options{'force'} eq "no"
	?
        ()
	:
	{
	    command    => $ENV{'GALACTICUS_EXEC_PATH'}."/Galacticus.exe ".$_->{'fileParameters'}." ".$ENV{'GALACTICUS_EXEC_PATH'}."/constraints/pipelines/darkMatter/changesTrue.xml",
	    launchFile => $_->{'filePrefix'}."_true.sh" ,
	    logFile    => $_->{'filePrefix'}."_true.log",
	    label      => "haloMassFunction_".$_->{'suite'}->{'name'}."_".$_->{'group'}->{'name'}."_".$_->{'simulation'}->{'name'}."_".$_->{'realization'}."_z".$_->{'redshift'}."_true",
	    ppn        => 1,
	    nodes      => 1,
	    mpi        => "no"
	}
    }
    @entries
    );
## Jobs using the Despali et al. (2015) halo mass function.
push(
    @jobs,
    map
    {
	-e $_->{'filePrefix'}.".hdf5_Despali2015:MPI0000" && $options{'force'} eq "no"
	?
        ()
	:
	{
	    command    => $ENV{'GALACTICUS_EXEC_PATH'}."/Galacticus.exe ".$_->{'fileParameters'}." ".$ENV{'GALACTICUS_EXEC_PATH'}."/constraints/pipelines/darkMatter/changesDespali2015.xml",
	    launchFile => $_->{'filePrefix'}."_despali2015.sh" ,
	    logFile    => $_->{'filePrefix'}."_despali2015.log",
	    label      => "haloMassFunction_".$_->{'suite'}->{'name'}."_".$_->{'group'}->{'name'}."_".$_->{'simulation'}->{'name'}."_".$_->{'realization'}."_z".$_->{'redshift'}."_despali2015",
	    ppn        => 1,
	    nodes      => 1,
	    mpi        => "no"
	}
    }
    @entries
    );
## Jobs using the Bohr et al. (2021) halo mass function.
push(
    @jobs,
    map
    {
	-e $_->{'filePrefix'}.".hdf5_Bohr2021:MPI0000" && $options{'force'} eq "no"
	?
        ()
	:
	{
	    command    => $ENV{'GALACTICUS_EXEC_PATH'}."/Galacticus.exe ".$_->{'fileParameters'}." ".$ENV{'GALACTICUS_EXEC_PATH'}."/constraints/pipelines/darkMatter/changesBohr2021.xml",
	    launchFile => $_->{'filePrefix'}."_bohr2021.sh" ,
	    logFile    => $_->{'filePrefix'}."_bohr2021.log",
	    label      => "haloMassFunction_".$_->{'suite'}->{'name'}."_".$_->{'group'}->{'name'}."_".$_->{'simulation'}->{'name'}."_".$_->{'realization'}."_z".$_->{'redshift'}."_bohr2021",
	    ppn        => 1,
	    nodes      => 1,
	    mpi        => "no"
	}
    }
    @entries
    );
&{$Galacticus::Launch::Hooks::moduleHooks{$queueManager->{'manager'}}->{'jobArrayLaunch'}}(\%options,@jobs)
    if ( scalar(@jobs) > 0 );

# Create plots of each individual model.
foreach my $entry ( @entries ) {
    next
	if ( -e $entry->{'filePrefix'}.".pdf" && $options{'force'} eq "no" );
    # Read the model halo mass function.
    my $dataModel;
    my $fileModel = new PDL::IO::HDF5($entry->{'filePrefix'}.":MPI0000.hdf5");
    my $output    = $fileModel->group('Outputs/Output1');
    $dataModel->{$_} = $output->dataset($_)->get()
	foreach ( 'haloMass', 'haloMassFunctionLnMBinAveraged' );
    $dataModel->{'nonZero'} = which($dataModel->{'haloMassFunctionLnMBinAveraged'} > 0);
    # Determine the minimum halo mass that was used in the fit.
    my $massHaloMinimum = $options{'countParticlesMinimum'}*$entry->{'group'}->{'massParticle'};
    # Read the target data.
    my $dataTarget;
    my $fileTarget       = new PDL::IO::HDF5($entry->{'fileTargetData'});
    my $targetSimulation = $fileTarget->group('simulation0001');
    $dataTarget->{$_} = $targetSimulation->dataset($_)->get()
	foreach ( 'mass', 'massFunction', 'count' );
    # Construct the target dataset errors.
    $dataTarget->{'massFunctionError'} = $dataTarget->{'massFunction'}->copy();
    $dataTarget->{'nonZero'          } = which(($dataTarget->{'count'} > 0) & ($dataTarget->{'mass'} >= $massHaloMinimum));
    $dataTarget->{'nonFit'           } = which(($dataTarget->{'count'} > 0) & ($dataTarget->{'mass'} <  $massHaloMinimum));
    $dataTarget->{'massFunctionError'}->($dataTarget->{'nonZero'}) /= $dataTarget->{'count'}->double()->($dataTarget->{'nonZero'})->sqrt();
    # Read the true model halo mass function.
    my $dataTrue;
    my $fileTrue   = new PDL::IO::HDF5($entry->{'filePrefix'}.".hdf5_true:MPI0000");
    my $outputTrue = $fileTrue->group('Outputs/Output1');
    $dataTrue->{$_} = $outputTrue->dataset($_)->get()
	foreach ( 'haloMass', 'haloMassFunctionLnMBinAveraged' );
    $dataTrue->{'nonZero'} = which($dataTrue->{'haloMassFunctionLnMBinAveraged'} > 0);
    # Read the Despali et al. (2015) halo mass function.
    my $dataDespali2015;
    my $fileDespali2015   = new PDL::IO::HDF5($entry->{'filePrefix'}.".hdf5_Despali2015:MPI0000");
    my $outputDespali2015 = $fileDespali2015->group('Outputs/Output1');
    $dataDespali2015->{$_} = $outputDespali2015->dataset($_)->get()
	foreach ( 'haloMass', 'haloMassFunctionLnMBinAveraged' );
    $dataDespali2015->{'nonZero'} = which($dataDespali2015->{'haloMassFunctionLnMBinAveraged'} > 0);
    # Read the Bohr et al. (2021) halo mass function.
    my $dataBohr2021;
    my $fileBohr2021   = new PDL::IO::HDF5($entry->{'filePrefix'}.".hdf5_Bohr2021:MPI0000");
    my $outputBohr2021 = $fileBohr2021->group('Outputs/Output1');
    $dataBohr2021->{$_} = $outputBohr2021->dataset($_)->get()
	foreach ( 'haloMass', 'haloMassFunctionLnMBinAveraged' );
    $dataBohr2021->{'nonZero'} = which($dataBohr2021->{'haloMassFunctionLnMBinAveraged'} > 0);
    # Create the plot.
    my $plot;
    my $gnuPlot;
    my $plotFileTeX = $entry->{'filePrefix'}.".tex";
    my $xMinimum = sprintf("%7.1e",10.0**floor($dataTarget->{'mass'        }->($dataTarget->{'nonZero'})->minimum()->log10()));
    my $xMaximum = sprintf("%7.1e",10.0**ceil ($dataTarget->{'mass'        }->($dataTarget->{'nonZero'})->maximum()->log10()));
    my $yMinimum = sprintf("%7.1e",10.0**floor($dataTarget->{'massFunction'}->($dataTarget->{'nonZero'})->minimum()->log10()));
    my $yMaximum = sprintf("%7.1e",10.0**ceil ($dataTarget->{'massFunction'}->($dataTarget->{'nonZero'})->maximum()->log10()));
    open($gnuPlot,"|gnuplot");
    print $gnuPlot "set terminal cairolatex pdf standalone color lw 2 size 4in,4in\n";
    print $gnuPlot "set output '".$plotFileTeX."'\n";
    print $gnuPlot "set title offset 0.0,-0.5 '".$entry->{'suite'}->{'name'}." ".$entry->{'group'}->{'name'}." ".$entry->{'simulation'}->{'name'}." ".$entry->{'realization'}." \$z=".$entry->{'redshift'}."\$'\n";
    print $gnuPlot "set xlabel '\$ M \$ [\$\\mathrm{M}_\\odot\$]'\n";
    print $gnuPlot "set ylabel '\$ \\mathrm{d} n / \\mathrm{d} \\log M \$ [Mpc\$^{-3}\$]\n";
    print $gnuPlot "set lmargin screen 0.15\n";
    print $gnuPlot "set rmargin screen 0.95\n";
    print $gnuPlot "set bmargin screen 0.15\n";
    print $gnuPlot "set tmargin screen 0.95\n";
    print $gnuPlot "set key spacing 1.2\n";
    print $gnuPlot "set logscale xy\n";
    print $gnuPlot "set mxtics 10\n";
    print $gnuPlot "set mytics 10\n";
    print $gnuPlot "set format x '\$10^{\%L}\$'\n";
    print $gnuPlot "set format y '\$10^{\%L}\$'\n";
    print $gnuPlot "set xrange [".$xMinimum.":".$xMaximum."]\n";
    print $gnuPlot "set yrange [".$yMinimum.":".$yMaximum."]\n";
    print $gnuPlot "set pointsize 1.0\n";
    my $lineX = pdl [ $massHaloMinimum, $massHaloMinimum ];
    my $lineY = pdl [ $yMinimum       , $yMaximum        ];
    &GnuPlot::PrettyPlots::Prepare_Dataset(
	\$plot                                                        ,
	$lineX                                                        ,
	$lineY                                                        ,
	style        => "line"                                        ,
	weight       => [2,1]                                         ,
	linePattern  => 2                                             ,
	color        => $GnuPlot::PrettyPlots::colorPairs{'blackGray'}
	);   
    &GnuPlot::PrettyPlots::Prepare_Dataset(
	\$plot                                                        ,
	$dataTarget->{'mass'        }->($dataTarget->{'nonFit'}),
	$dataTarget->{'massFunction'}->($dataTarget->{'nonFit'}),
	errorDown    => $dataTarget->{'massFunctionError'}->($dataTarget->{'nonZero'}),
	errorUp      => $dataTarget->{'massFunctionError'}->($dataTarget->{'nonZero'}),
	style        => "point"                                        ,
	symbol       => [6,7]                                         ,
	weight       => [2,1]                                         ,
	pointSize    => 0.25                                             ,
	transparency => 0.9,
	color        => $GnuPlot::PrettyPlots::colorPairs{'mediumSeaGreen'}
	);   
    &GnuPlot::PrettyPlots::Prepare_Dataset(
	\$plot                                                        ,
	$dataTarget->{'mass'        }->($dataTarget->{'nonZero'}),
	$dataTarget->{'massFunction'}->($dataTarget->{'nonZero'}),
	errorDown    => $dataTarget->{'massFunctionError'}->($dataTarget->{'nonZero'}),
	errorUp      => $dataTarget->{'massFunctionError'}->($dataTarget->{'nonZero'}),
	style        => "point"                                          ,
	symbol       => [6,7]                                            ,
	weight       => [2,1]                                            ,
	pointSize    => 0.25                                             ,
	color        => $GnuPlot::PrettyPlots::colorPairs{'mediumSeaGreen'}
	);   
    &GnuPlot::PrettyPlots::Prepare_Dataset(
	\$plot                                                                               ,
	$dataDespali2015->{'haloMass'                      }->($dataDespali2015->{'nonZero'}),
	$dataDespali2015->{'haloMassFunctionLnMBinAveraged'}->($dataDespali2015->{'nonZero'}),
	style        => "line"                                                               ,
	weight       => [2,1]                                                                ,
	linePattern  => 3,
	color        => $GnuPlot::PrettyPlots::colorPairs{'cornflowerBlue'}
	);   
    &GnuPlot::PrettyPlots::Prepare_Dataset(
	\$plot                                                                         ,
	$dataBohr2021->{'haloMass'                      }->($dataBohr2021->{'nonZero'}),
	$dataBohr2021->{'haloMassFunctionLnMBinAveraged'}->($dataBohr2021->{'nonZero'}),
	style        => "line"                                                         ,
	weight       => [2,1]                                                          ,
	linePattern  => 3,
	color        => $GnuPlot::PrettyPlots::colorPairs{'lightGray'}
	);   
    &GnuPlot::PrettyPlots::Prepare_Dataset(
	\$plot                                                                 ,
	$dataTrue->{'haloMass'                      }->($dataTrue->{'nonZero'}),
	$dataTrue->{'haloMassFunctionLnMBinAveraged'}->($dataTrue->{'nonZero'}),
	style        => "line"                                                 ,
	weight       => [2,1]                                                  ,
	linePattern  => 2,
	color        => $GnuPlot::PrettyPlots::colorPairs{'redYellow'}
	);   
    &GnuPlot::PrettyPlots::Prepare_Dataset(
	\$plot                                                                   ,
	$dataModel->{'haloMass'                      }->($dataModel->{'nonZero'}),
	$dataModel->{'haloMassFunctionLnMBinAveraged'}->($dataModel->{'nonZero'}),
	style        => "line"                                                   ,
	weight       => [2,1]                                                    ,
	color        => $GnuPlot::PrettyPlots::colorPairs{'redYellow'}
	);   
    &GnuPlot::PrettyPlots::Plot_Datasets($gnuPlot,\$plot);
    close($gnuPlot);
    &GnuPlot::LaTeX::GnuPlot2PDF($plotFileTeX);
}

exit 0;

File: constraints/pipelines/darkMatter/haloMassFunctionPostProcess.py
#!/usr/bin/env python3
"""Script to generate post-processed halo mass function models and plots.

Python port of constraints/pipelines/darkMatter/haloMassFunctionPostProcess.pl
Andrew Benson (ported to Python 2026)
"""

import argparse
import os

import h5py
import matplotlib
matplotlib.use('Agg')
import matplotlib.pyplot as plt
import numpy as np

import queueManager
from queueManager import translate_job, submit_jobs
from Galacticus.Constraints.Simulations import iterate, parse_simulations_xml
from Galacticus._logging                 import configure_default as _configure_default

# Show INFO-level diagnostic output from library modules.
_configure_default()


# ---------------------------------------------------------------------------
# Argument parsing
# ---------------------------------------------------------------------------

def _parse_args():
    parser = argparse.ArgumentParser(
        description='Generate post-processed halo mass function models and plots.'
    )
    parser.add_argument('--pipelinePath',           required=True,
                        help='Path to pipeline configuration files')
    parser.add_argument('--outputDirectory',        required=True,
                        help='Directory for generated output files')
    parser.add_argument('--force',                  default='no',
                        help='Re-run models even if outputs exist (default: no)')
    parser.add_argument('--countParticlesMinimum',  default=300, type=int,
                        help='Minimum particles per halo for constraints (default: 300)')
    parser.add_argument('--select',                 default=None, action='append',
                        help='Simulation selection filter (suite::group::resolution::...); may be repeated')
    parser.add_argument('--partition',              default=None,
                        help='SLURM partition override')
    parser.add_argument('--jobMaximum',             default=None, type=int,
                        help='Maximum number of concurrent jobs')
    parser.add_argument('--waitOnSubmit',       default=None, type=int,
                        help='Time (in seconds) to wait after submitting')
    parser.add_argument('--waitOnActive',       default=None, type=int,
                        help='Time (in seconds) to wait between polling active jobs')
    args = parser.parse_args()
    for key in ('pipelinePath', 'outputDirectory'):
        val = getattr(args, key)
        if val and not val.endswith('/'):
            setattr(args, key, val + '/')
    return args


# ---------------------------------------------------------------------------
# Job construction
# ---------------------------------------------------------------------------

def _build_jobs(entries, options):
    """Return list of Perl-style job dicts for the four model variants."""
    galacticus_exe = os.environ['GALACTICUS_EXEC_PATH'] + '/Galacticus.exe'
    pipeline_path  = os.environ['GALACTICUS_EXEC_PATH'] + '/constraints/pipelines/darkMatter/'
    force          = options['force']
    jobs           = []

    variants = [
        # (suffix, extra_xml, output_check_suffix)
        ('',           None,                    ':MPI0000.hdf5'           ),
        ('_true',      'changesTrue.xml',        '.hdf5_true:MPI0000'      ),
        ('_despali2015', 'changesDespali2015.xml', '.hdf5_Despali2015:MPI0000'),
        ('_bohr2021',  'changesBohr2021.xml',    '.hdf5_Bohr2021:MPI0000'  ),
    ]

    for entry in entries:
        file_prefix = entry['filePrefix']
        file_params = entry['fileParameters']
        label_base  = (
            f"haloMassFunction_{entry['suite']['name']}_{entry['group']['name']}_"
            f"{entry['resolution']['name']}_{entry['simulation']['name']}_"
            f"{entry['realization']}_z{entry['redshift']}"
        )

        for suffix, changes_xml, check_suffix in variants:
            output_file = file_prefix + check_suffix
            if os.path.exists(output_file) and force == 'no':
                continue
            command = f'{galacticus_exe} {file_params}'
            if changes_xml is not None:
                command += f' {pipeline_path}{changes_xml}'
            jobs.append({
                'command':    command,
                'launchFile': f'{file_prefix}{suffix}.sh',
                'logFile':    f'{file_prefix}{suffix}.log',
                'label':      f'{label_base}{suffix}',
                'ppn':        1,
                'nodes':      1,
                'walltime':   '2:00:00',
                'mpi':        'no',
            })

    return jobs


# ---------------------------------------------------------------------------
# Plotting
# ---------------------------------------------------------------------------

def _read_hmf(hdf5_path):
    """Read haloMass and haloMassFunctionLnMBinAveraged from Outputs/Output1."""
    with h5py.File(hdf5_path, 'r') as f:
        grp = f['Outputs/Output1']
        mass = grp['haloMass'][:]
        hmf  = grp['haloMassFunctionLnMBinAveraged'][:]
    return mass, hmf


def _plot_entry(entry, options):
    file_prefix  = entry['filePrefix']
    output_pdf   = file_prefix + '.pdf'
    if os.path.exists(output_pdf) and options['force'] == 'no':
        return

    # Read model outputs.
    mass_model,       hmf_model       = _read_hmf(file_prefix + ':MPI0000.hdf5')
    mass_true,        hmf_true        = _read_hmf(file_prefix + '.hdf5_true:MPI0000')
    mass_despali2015, hmf_despali2015 = _read_hmf(file_prefix + '.hdf5_Despali2015:MPI0000')
    mass_bohr2021,    hmf_bohr2021    = _read_hmf(file_prefix + '.hdf5_Bohr2021:MPI0000')

    # Read target data.
    with h5py.File(entry['fileTargetData'], 'r') as f:
        grp          = f['simulation0001']
        mass_target  = grp['mass'][:]
        hmf_target   = grp['massFunction'][:]
        count_target = grp['count'][:]

    # Derive minimum halo mass threshold and filter masks.
    mass_halo_min   = options['countParticlesMinimum'] * entry['resolution']['massParticle']
    hmf_err         = hmf_target.copy()
    non_zero        = np.where((count_target > 0) & (mass_target >= mass_halo_min))[0]
    non_fit         = np.where((count_target > 0) & (mass_target <  mass_halo_min))[0]
    hmf_err[non_zero] /= np.sqrt(count_target[non_zero].astype(float))

    # Determine axis ranges (decade-rounded, matching Perl floor/ceil logic).
    x_min = 10 ** np.floor(np.log10(mass_target[non_zero].min()))
    x_max = 10 ** np.ceil( np.log10(mass_target[non_zero].max()))
    y_min = 10 ** np.floor(np.log10(hmf_target[non_zero].min()))
    y_max = 10 ** np.ceil( np.log10(hmf_target[non_zero].max()))

    # Build the figure.
    fig, ax = plt.subplots(figsize=(4, 4))
    ax.set_xscale('log')
    ax.set_yscale('log')
    ax.set_xlim(x_min, x_max)
    ax.set_ylim(y_min, y_max)
    ax.set_xlabel(r'$M$ [$\mathrm{M}_\odot$]')
    ax.set_ylabel(r'$\mathrm{d}n / \mathrm{d}\log M$ [Mpc$^{-3}$]')
    ax.set_title(
        f"{entry['suite']['name']} {entry['group']['name']} "
        f"{entry['resolution']['name']} {entry['simulation']['name']} "
        f"{entry['realization']} $z={entry['redshift']}$",
        fontsize=8
    )

    # 1. Vertical line at minimum halo mass.
    ax.axvline(mass_halo_min, color='dimgray', linestyle='--', linewidth=1)

    # 2. Target data below fit threshold (transparent).
    if non_fit.size > 0:
        ax.errorbar(
            mass_target[non_fit], hmf_target[non_fit],
            yerr=hmf_err[non_fit],
            fmt='o', markersize=2, color='mediumseagreen', alpha=0.25,
            elinewidth=0.5, capsize=0
        )

    # 3. Target data in fit region.
    if non_zero.size > 0:
        ax.errorbar(
            mass_target[non_zero], hmf_target[non_zero],
            yerr=hmf_err[non_zero],
            fmt='o', markersize=2, color='mediumseagreen',
            elinewidth=0.5, capsize=0
        )

    # 4. Despali et al. (2015) model.
    nz = hmf_despali2015 > 0
    if nz.any():
        ax.plot(mass_despali2015[nz], hmf_despali2015[nz],
                color='cornflowerblue', linestyle='--', linewidth=1)

    # 5. Bohr et al. (2021) model.
    nz = hmf_bohr2021 > 0
    if nz.any():
        ax.plot(mass_bohr2021[nz], hmf_bohr2021[nz],
                color='lightgray', linestyle='--', linewidth=1)

    # 6. True model (no numerical artifacts).
    nz = hmf_true > 0
    if nz.any():
        ax.plot(mass_true[nz], hmf_true[nz],
                color='orangered', linestyle='--', linewidth=1)

    # 7. Base (fitted) model.
    nz = hmf_model > 0
    if nz.any():
        ax.plot(mass_model[nz], hmf_model[nz],
                color='orangered', linestyle='-', linewidth=1)

    fig.savefig(output_pdf, bbox_inches='tight')
    plt.close(fig)


# ---------------------------------------------------------------------------
# main
# ---------------------------------------------------------------------------

def main():
    args    = _parse_args()
    options = vars(args)

    simulations = parse_simulations_xml(options['pipelinePath'] + 'simulations.xml')

    # Perl: force re-run when --select is given.
    if options.get('select') is not None:
        options['force'] = 'yes'

    # Build full entry list and attach derived file paths.
    entries = iterate(simulations, options)
    for entry in entries:
        identifier = (
            f"{entry['suite']['name']}_{entry['group']['name']}_"
            f"{entry['resolution']['name']}_{entry['simulation']['name']}_"
            f"{entry['realization']}_z{entry['redshift']}"
        )
        entry['fileParameters'] = options['outputDirectory'] + f'haloMassFunctionBase_{identifier}.xml'
        entry['filePrefix']     = options['outputDirectory'] + f'haloMassFunction_{identifier}'

    # Group by (suite, group, resolution, simulation, redshift) — mirrors Perl %setsRealization.
    sets_realization = {}
    for entry in entries:
        key = (
            f"{entry['suite']['name']}_{entry['group']['name']}_"
            f"{entry['resolution']['name']}_{entry['simulation']['name']}_"
            f"z{entry['redshift']}"
        )
        sets_realization.setdefault(key, []).append(entry)

    # Submit Galacticus jobs for all four model variants.
    jobs = _build_jobs(entries, options)
    if jobs:
        manager = queueManager.factory(args)
        submit_jobs(manager, jobs)

    # Generate comparison plots.
    for entry in entries:
        _plot_entry(entry, options)


if __name__ == '__main__':
    main()
